builder/openstack/step_source_image_info.go

```diff
@@ -0,0 +1,96 @@
+package openstack
+
+import (
+	"context"
+	"fmt"
+	"log"
+
+	"github.com/gophercloud/gophercloud/openstack/imageservice/v2/images"
+	"github.com/gophercloud/gophercloud/pagination"
+	"github.com/hashicorp/packer/helper/multistep"
+	"github.com/hashicorp/packer/packer"
+)
+
+type StepSourceImageInfo struct {
+	SourceImage     string
+	SourceImageName string
+	ImageFilters    ImageFilterOptions
+}
+
+type ImageFilterOptions struct {
+	Filters    map[string]interface{} `mapstructure:"filters"`
+	MostRecent bool                   `mapstructure:"most_recent"`
+}
+
+func (s *StepSourceImageInfo) Run(_ context.Context, state multistep.StateBag) multistep.StepAction {
+	config := state.Get("config").(Config)
+	ui := state.Get("ui").(packer.Ui)
+
+	client, err := config.computeV2Client()
+
+	// if an ID is provided we skip the filter since that will return a single or no image
+	if s.SourceImage != "" {
+		state.Put("source_image", s.SourceImage)
+		return multistep.ActionContinue
+	}
+
+	params := &images.ListOpts{}
+
+	// build ListOpts from filters
+	if len(s.ImageFilters.Filters) > 0 {
+		err = buildImageFilters(s.ImageFilters.Filters, params)
+		if err != nil {
+			err := fmt.Errorf("Errors encountered in filter parsing.\n%s" + err.Error())
+			state.Put("error", err)
+			ui.Error(err.Error())
+			return multistep.ActionHalt
+		}
+	}
+
+	// override the "name" provided in filters if "s.SourceImageName" was filled
+	if s.SourceImageName != "" {
+		params.Name = s.SourceImageName
+	}
+
+	// apply "most_recent" logic to the sort fields and allow OpenStack to return the latest qualified image
+	if s.ImageFilters.MostRecent {
+		applyMostRecent(params)
+	}
+
+	log.Printf("Using Image Filters %v", params)
+	image := &images.Image{}
+	err = images.List(client, params).EachPage(func(page pagination.Page) (bool, error) {
+		i, err := images.ExtractImages(page)
+		if err != nil {
+			return false, err
+		}
+
+		switch len(i) {
+		case 0:
+			return false, fmt.Errorf("No image was found matching filters: %v", params)
+		case 1:
+			*image = i[0]
+			return true, nil
+		default:
+			return false, fmt.Errorf(
```

Comment:
Possible fix:
```
if s.ImageFilters.MostRecent {
	*image = i[0]
	return true, nil
}
return false, fmt.Errorf("Your query returned more than one result. Please try a more specific search, or set most_recent to true. Search filters: %v", params)

```

Reply:
Yes, I never set limit when doing the search. I was planning to have the gophercloud search limit set to one so we don't have to receive/process as much data. 

```go
func applyMostRecent(listOpts *images.ListOpts) {
	listOpts.Sort = mostRecentSort
	listOpts.Limit = 1 // <- update

	return
}
```

I'll be doing it that way unless you prefer it done on the Packer side

Reply:
I actually tried with the limit but I think I did not get the latest image then. Can try that again too see what result I got. That was the reason I put the if statement instead. 😄 Will test it again now.

Reply:
Just checked now and using `Limit` does NOT work as you would expect. The latest created image is not returned then.